File: patron/src/main/java/com/prolificinteractive/patrons/BasePreference.java
package com.prolificinteractive.patrons;

import android.content.SharedPreferences;

/**
 * The type Base preference.
 *
 * @param <T> the type parameter hold by the shared preference.
 */
public abstract class BasePreference<T> implements Preference<T> {
  /**
   * The shared preferences.
   */
  protected final SharedPreferences preferences;

  /**
   * The shared preference Key.
   */
  protected final String key;

  /**
   * The Default value.
   */
  protected final T defaultValue;

  /**
   * Instantiates a new Base preference.
   *
   * @param preferences the shared preferences
   * @param key the key
   */
  public BasePreference(final SharedPreferences preferences, final String key) {
    this(preferences, key, null);
  }

  /**
   * Instantiates a new Base preference.
   *
   * @param preferences the shared preferences
   * @param key the key
   * @param defaultValue the default value
   */
  public BasePreference(
      final SharedPreferences preferences,
      final String key,
      final T defaultValue) {
    this.preferences = preferences;
    this.key = key;
    this.defaultValue = defaultValue;
  }

  @Override public void delete() {
    preferences.edit().remove(key).apply();
  }

  @Override public boolean isSet() {
    return preferences.contains(key);
  }
}

File: patron/src/main/java/com/prolificinteractive/patrons/BasePreference.kt
package com.prolificinteractive.patrons

import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener

/**
 * The type Base preference.
 *
 * @param <T> the type parameter hold by the shared preference.
 */
/**
 * Instantiates a new Base preference.
 *
 * @param preferences the shared preferences
 * @param key the key
 * @param defaultValue the default value
 */
abstract class BasePreference<T>(
    /**
     * The shared preferences.
     */
    protected val preferences: SharedPreferences,
    /**
     * The shared preference Key.
     */
    protected val key: String,
    /**
     * The Default value.
     */
    protected val defaultValue: T
) : Preference<T> {

  private val listeners: MutableMap<OnPreferenceChangeListener<T?>, OnSharedPreferenceChangeListener> = mutableMapOf()

  override fun delete() = preferences.edit().remove(key).apply()

  override fun isSet() = preferences.contains(key)

  override fun registerChangeListener(listener: OnPreferenceChangeListener<T?>) {
    val l = OnSharedPreferenceChangeListener { _, _ -> listener(get()) }
    preferences.registerOnSharedPreferenceChangeListener(l)
    listeners[listener] = l
  }

  override fun unregisterChangeListener(listener: OnPreferenceChangeListener<T?>) {
    val l = listeners[listener]
    preferences.unregisterOnSharedPreferenceChangeListener(l)
    listeners.remove(listener)
  }

  override fun unregisterAllChangeListener() {
    listeners.forEach { preferences.unregisterOnSharedPreferenceChangeListener(it.value) }
    listeners.clear()
  }
}